exonum/src/sandbox/mod.rs

```diff
@@ -12,10 +12,1402 @@
+        });
+        network_getter.wait().unwrap();
+    }
+
+    fn process_internal_requests(&mut self) {
+        let internal_getter = futures::lazy(|| -> Result<(), ()> {
+            while let Async::Ready(Some(internal)) = self.internal_requests_rx.poll()? {
+                match internal {
+                    InternalRequest::Timeout(t) => self.timers.push(t),
+                    InternalRequest::JumpToRound(height, round) => self
+                        .handler
+                        .handle_event(InternalEvent::JumpToRound(height, round).into()),
+                    InternalRequest::Shutdown => unimplemented!(),
+                    InternalRequest::VerifyMessage(message) => {
+                        let protocol =
+                            Message::deserialize(SignedMessage::from_raw_buffer(message).unwrap())
+                                .unwrap();
+                        self.handler.handle_event(
+                            InternalEvent::MessageVerified(Box::new(protocol)).into(),
+                        );
+                    }
+                }
+            }
+            Ok(())
+        });
+        internal_getter.wait().unwrap();
+    }
+    fn process_api_requests(&mut self) {
+        let api_getter = futures::lazy(|| -> Result<(), ()> {
+            while let Async::Ready(Some(api)) = self.api_requests_rx.poll()? {
+                self.handler.handle_event(api.into());
+            }
+            Ok(())
+        });
+        api_getter.wait().unwrap();
+    }
+}
+
+pub struct Sandbox {
+    pub validators_map: HashMap<PublicKey, SecretKey>,
+    pub services_map: HashMap<PublicKey, SecretKey>,
+    inner: RefCell<SandboxInner>,
+    addresses: Vec<ConnectInfo>,
+    /// Connect message used during initialization.
+    connect: Option<Signed<Connect>>,
+}
+
+impl Sandbox {
+    pub fn initialize(
+        &mut self,
+        connect_message_time: SystemTime,
+        start_index: usize,
+        end_index: usize,
+    ) {
+        let connect = self.create_connect(
+            &self.p(ValidatorId(0)),
+            self.a(ValidatorId(0)),
+            connect_message_time.into(),
+            &user_agent::get(),
+            self.s(ValidatorId(0)),
+        );
+
+        for validator in start_index..end_index {
+            let validator = ValidatorId(validator as u16);
+            self.recv(&self.create_connect(
+                &self.p(validator),
+                self.a(validator),
+                self.time().into(),
+                &user_agent::get(),
+                self.s(validator),
+            ));
+            self.send(self.p(validator), &connect);
+        }
+
+        self.check_unexpected_message();
+        self.connect = Some(connect);
+    }
+
+    fn check_unexpected_message(&self) {
+        if let Some((addr, msg)) = self.pop_sent() {
+            panic!("Send unexpected message {:?} to {}", msg, addr);
+        }
+    }
+
+    pub fn p(&self, id: ValidatorId) -> PublicKey {
```

Comment:
I think since you began to refactor sandbox, you may rename this methods from `a` to `address` and from `p` to `pub_key` and from `s` to `sec_key` to clarity.

Reply:
Or just `addr`